File: RailgunVB/Core/Managers/FilterManager.vb
Imports Discord
Imports Microsoft.Extensions.DependencyInjection
Imports RailgunVB.Core.Filters
Imports TreeDiagram

Namespace Core.Managers
    
    Public Class FilterManager
    
        Private ReadOnly _filters As New List(Of IMessageFilter)()
        Private ReadOnly _services As IServiceProvider
        
        Public Sub New(services As IServiceProvider)
            _services = services
        End Sub
        
        Public Sub RegisterFilter(filter As IMessageFilter)
            _filters.Add(filter)
        End Sub
        
        Public Async Function ApplyFilterAsync(msg As IUserMessage) As Task(Of IUserMessage)
            Dim result As IUserMessage = Nothing
            
            Using scope As IServiceScope = _services.CreateScope()
                Dim db As TreeDiagramContext = scope.ServiceProvider.GetService(Of TreeDiagramContext)
                For Each filter As IMessageFilter In _filters
                    result = Await filter.FilterAsync(msg, db)
                
                    If result IsNot Nothing Then Exit For
                Next
            End Using
            
            Return result
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Managers/FilterManager.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Discord;
using Microsoft.Extensions.DependencyInjection;
using Railgun.Core.Filters;
using TreeDiagram;

namespace Railgun.Core.Managers
{
    public class FilterManager
    {
        private readonly IServiceProvider _services;
        private readonly List<IMessageFilter> _filters = new List<IMessageFilter>();

        public FilterManager(IServiceProvider services)
            => _services = services;

        public void AddMessageFilter(IMessageFilter filter) => _filters.Add(filter);

        public async Task<IUserMessage> ApplyFilterAsync(IUserMessage msg) {
            if (string.IsNullOrWhiteSpace(msg.Content)) return null;

            var tc = (ITextChannel)msg.Channel;
            IUserMessage result = null;

            using (var scope = _services.CreateScope()) {
                var db = scope.ServiceProvider.GetService<TreeDiagramContext>();

                foreach (var filter in _filters) {
                    result = await filter.FilterAsync(tc, msg, db);

                    if (result != null) break;
                }
            }

            return result;
        }
    }
}